xds/internal/xdsclient/client_new.go

```diff
@@ -124,56 +115,69 @@ func triggerXDSResourceNameNotFoundClient(resourceType, resourceName string) err
+func NewForTesting(opts OptionsForTesting) (XDSClient, func(), error) {
+	if opts.WatchExpiryTimeout == 0 {
+		opts.WatchExpiryTimeout = defaultWatchExpiryTimeout
+	}
+	if opts.AuthorityIdleTimeout == 0 {
+		opts.AuthorityIdleTimeout = defaultIdleAuthorityDeleteTimeout
+	}
+
 	// Normalize the contents
 	buf := bytes.Buffer{}
-	err := json.Indent(&buf, contents, "", "")
+	err := json.Indent(&buf, opts.Contents, "", "")
 	if err != nil {
 		return nil, nil, fmt.Errorf("xds: error normalizing JSON: %v", err)
 	}
-	contents = bytes.TrimSpace(buf.Bytes())
+	opts.Contents = bytes.TrimSpace(buf.Bytes())
 
-	c, err := getOrMakeClientForTesting(contents)
-	if err != nil {
-		return nil, nil, err
-	}
-	singletonClientForTesting.Store(c)
-	return c, grpcsync.OnceFunc(func() {
+	clientsMu.Lock()
+	defer clientsMu.Unlock()
+
+	var client *clientRefCounted
+	closeFunc := grpcsync.OnceFunc(func() {
 		clientsMu.Lock()
 		defer clientsMu.Unlock()
-		if c.decrRef() == 0 {
-			c.close()
-			delete(clients, string(contents))
-			singletonClientForTesting.Store(nil)
+		if client.decrRef() == 0 {
+			client.close()
+			delete(clients, string(opts.Contents))
 		}
-	}), nil
-}
+	})
 
-// getOrMakeClientForTesting creates a new reference counted client (separate
-// from the global singleton) for the given config, or returns an existing one.
-// It takes care of incrementing the reference count for the returned client,
-// and leaves the caller responsible for decrementing the reference count once
-// the client is no longer needed.
-func getOrMakeClientForTesting(config []byte) (*clientRefCounted, error) {
-	clientsMu.Lock()
-	defer clientsMu.Unlock()
-
-	if c := clients[string(config)]; c != nil {
+	// If an xDS client exists for the given configuration, increment its
+	// reference count and return it.
+	if c := clients[string(opts.Contents)]; c != nil {
 		c.incrRef()
-		return c, nil
+		client = c
+		return c, closeFunc, nil
 	}
 
-	bcfg, err := bootstrap.NewConfigFromContents(config)
+	// Create a new xDS client for the given configuration
+	bcfg, err := bootstrap.NewConfigFromContents(opts.Contents)
 	if err != nil {
-		return nil, fmt.Errorf("bootstrap config %s: %v", string(config), err)
+		return nil, nil, fmt.Errorf("bootstrap config %s: %v", string(opts.Contents), err)
 	}
-	cImpl, err := newWithConfig(bcfg, defaultWatchExpiryTimeout, defaultIdleAuthorityDeleteTimeout)
+	cImpl, err := newWithConfig(bcfg, opts.WatchExpiryTimeout, opts.AuthorityIdleTimeout)
 	if err != nil {
-		return nil, fmt.Errorf("creating xDS client: %v", err)
+		return nil, nil, fmt.Errorf("creating xDS client: %v", err)
 	}
 	c := &clientRefCounted{clientImpl: cImpl, refCount: 1}
-	clients[string(config)] = c
-	return c, nil
+	clients[string(opts.Contents)] = c
+	client = c
+
+	return c, closeFunc, nil
+}
+
+func init() {
+	internal.TriggerXDSResourceNotFoundForTesting = triggerXDSResourceNotFoundForTesting
```

Comment:
This is a bit unfortunate...if we're going to have tests for xds living outside of `xds/` then is that another reason to move all of `xds/internal` into `internal/xds`?  Because we could just export this function directly instead of doing all the dance to trampoline through `internal` and losing static type safety.

We even still could expose it for use from inside `xds/`...